x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -338,14 +337,20 @@ private void createApiKeyAndIndexIt(
                         SECURITY_ORIGIN,
                         BulkAction.INSTANCE,
                         bulkRequest,
-                        TransportSingleItemBulkWriteAction.<IndexResponse>wrapBulkResponse(ActionListener.wrap(indexResponse -> {
-                            assert request.getId().equals(indexResponse.getId());
-                            assert indexResponse.getResult() == DocWriteResponse.Result.CREATED;
-                            final ListenableFuture<CachedApiKeyHashResult> listenableFuture = new ListenableFuture<>();
-                            listenableFuture.onResponse(new CachedApiKeyHashResult(true, apiKey));
-                            apiKeyAuthCache.put(request.getId(), listenableFuture);
-                            listener.onResponse(new CreateApiKeyResponse(request.getName(), request.getId(), apiKey, expiration));
-                        }, listener::onFailure))
+                        ActionListener.wrap(bulkResponse -> {
+                            assert bulkResponse.getItems().length == 1;
+                            final BulkItemResponse indexResponse = bulkResponse.getItems()[0];
+                            if (indexResponse.isFailed()) {
+                                listener.onFailure(indexResponse.getFailure().getCause());
+                            } else {
+                                assert request.getId().equals(indexResponse.getResponse().getId());
+                                assert indexResponse.getResponse().getResult() == DocWriteResponse.Result.CREATED;
+                                final ListenableFuture<CachedApiKeyHashResult> listenableFuture = new ListenableFuture<>();
+                                listenableFuture.onResponse(new CachedApiKeyHashResult(true, apiKey));
+                                apiKeyAuthCache.put(request.getId(), listenableFuture);
+                                listener.onResponse(new CreateApiKeyResponse(request.getName(), request.getId(), apiKey, expiration));
+                            }
+                        }, listener::onFailure)
```

Comment:
I meant to comment on this but somehow forgot. It is for future refactoring: This `wrapBulkResponse` method is used in a few other places. It might be helpful to move this method to a different non-deprecated class (TransportBulkAction?) so that we can keep reusing it instead of having to repeat the same logic in other palces?

Reply:
Absolutely, was thinking this too. I'll open a separate PR with a util method. Is this best reviewed by the distributed team or core or another team I'm not thinking of?

Reply:
I think technically it should be the distributed team.